modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java

```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
             }
         }
     }
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutWorker extends IgniteWorker {
+        /** Worker sleep interval. */
+        private final long sleepInterval = getLong("IGNITE_TIMEOUT_WORKER_SLEEP_INTERVAL", 500);
+
+        /** Active operation. */
+        public final ConcurrentMap<Long, TimeoutObject> requestsMap;
+
+        /**
+         * Constructor.
+         *
+         * @param log Logger.
+         * @param igniteInstanceName Name of the Ignite instance this runnable is used in.
+         * @param name Worker name. Note that in general thread name and worker (runnable) name are two different things. The same
+         *         worker can be executed by multiple threads and therefore for logging and debugging purposes we separate the two.
+         * @param requestsMap Active operations.
+         */
+        public TimeoutWorker(
+                IgniteLogger log,
+                String igniteInstanceName,
+                String name,
+                ConcurrentMap<Long, TimeoutObject> requestsMap
+        ) {
+            super(log, igniteInstanceName, name, null);
+
+            this.requestsMap = requestsMap;
+        }
+
+        @Override
+        protected void body() {
+            try {
+                TimeoutObject timeoutObject;
+
+                while (!isCancelled()) {
+                    long now = coarseCurrentTimeMillis();
+
+                    for (Entry<Long, TimeoutObject> entry : new HashMap<>(requestsMap).entrySet()) {
+                        updateHeartbeat();
+
+                        timeoutObject = entry.getValue();
+
+                        assert timeoutObject != null : "Unexpected null on the timeout queue.";
+
+                        if (timeoutObject.getEndTime() > 0 && now > timeoutObject.getEndTime()) {
+                            CompletableFuture<?> fut = timeoutObject.getFuture();
+
+                            if (!fut.isDone()) {
```

Comment:
Is this check really needed? If the future is already completed, its subsequent completion will have no effect, and it will take just one volatile read (which we pay for `isDone()`)

Reply:
But we would have to create exception instances. This code is not in a low-latency path, and I was worried about GC here.